ts2phc.8
.TH TS2PHC 8 "March 2024" "linuxptp"
.SH NAME
ts2phc - Synchronizes one or more PTP Hardware Clocks using external time stamps.

.SH SYNOPSIS
.B ts2phc
[
.B \-hmqv
] [
.BI \-c " device|name"
] [
.BI \-f " config"
] [
.BI \-l " print-level"
] [
.BI \-s " device|name"
] [
.I long-options
]
.I .\|.\|.

.SH DESCRIPTION
.B ts2phc
synchronizes PTP Hardware Clocks (PHC) to external time stamp signals.
A single source may be used to distribute time to one or more PHC devices.

.SH OPTIONS
.TP
.BI \-a
Adjust the direction of synchronization automatically. The program determines
which PHC should the reference clock for time distribution and which should
be the destinations by querying the port states from the running instance of
.B ptp4l.
Note that using this option, the PPS signal distribution hierarchy still
remains fixed as per the configuration file. This implies that using this
option, a PPS source of the PHC kind may become a target clock, and a PPS sink
may become a reference clock. Other, non-PHC types of PPS sources (generic,
NMEA) cannot become target clocks. Clocks which are not part of
.B ptp4l's
list of ports are not synchronized. This option is useful when the
.B boundary_clock_jbod
option of ptp4l is also enabled.
.TP
.BI \-c " device|name"
Specifies a PHC to be synchronized.
The clock may be identified by its character device (like /dev/ptp0)
or its associated network interface (like eth0).
This option may be given multiple times.
.TP
.BI \-f " config"
Read configuration from the specified file.
No configuration file is read by default.
.TP
.BI \-h
Displays the command line help summary.
.TP
.BI \-l " print-level"
Sets the maximum syslog level of messages which should be printed or
sent to the system logger. The default is 6 (LOG_INFO).
.TP
.B \-m
Prints log messages to the standard output.
.TP
.B \-q
Prevents sending log messages to the system logger.
.TP
.BI \-s " device|name"
Specifies the source of the Time of Day (ToD) data.
Use the key word "generic" for an external 1-PPS without ToD information.
When using a PHC as the time source, the clock may be identified by its character
device (like /dev/ptp0) or its associated network interface (like
eth0).
Use the key word "nmea" for an external 1-PPS from a GPS providing ToD
information via the RMC NMEA sentence.
.TP
.B \-v
Prints the software version and exits.

.SH LONG OPTIONS

Each and every configuration file option (see below) may also appear
as a "long" style command line argument.  For example, the use_syslog
option may be set using either of these two forms.

.RS
\f(CW\-\-use_syslog 1   \-\-use_syslog=1\fP
.RE

Option values given on the command line override values in the global
section of the configuration file.

.SH CONFIGURATION FILE

The configuration file is divided into sections. Each section starts with a
line containing its name enclosed in brackets and it follows with settings.
Each setting is placed on a separate line, it contains the name of the
option and the value separated by whitespace characters. Empty lines and lines
starting with # are ignored.

There are two different section types.

.TP
.B 1.
The global section (indicated as
.BR [global] )
sets the program options and the default time sink options. Other
sections are clock specific sections, and they override the default
options.
.TP
.B 2.
Time sink sections give the name of the configured PHC (e.g.
.BR [eth0] ).
Time sinks specified in the configuration file need not be specified
with the
.B \-c
command line option.

.SH GLOBAL OPTIONS

.TP
.B active_key_id
Used in conjunction with \fBspp\fR and \fBsa_file\fR directives to
specify which key from the \fBspp\fR defined Security Association
should be used for outbound icv calculations. All Security Associations
are read from the file specified by \fBsa_file\fR. Requires \fBspp\fR
and \fBsa_file\fR directives. Must be in the range of 1 to 2^32-1,
inclusive. The default is 0 (disabled).

.TP
.B clock_servo
The servo which is used to synchronize the local clock. Valid values
are "pi" for a PI controller, "linreg" for an adaptive controller
using linear regression, "ntpshm" and "refclock_sock" for the NTP SHM and
chrony SOCK reference clocks respectively to allow another process to
synchronize the local clock, and "nullf" for a servo that always dials
frequency offset zero (for use in SyncE nodes).
The default is "pi".

.TP
.B first_step_threshold
The maximum offset, specified in seconds, that the servo will correct by
changing the clock frequency (phase when using nullf servo) instead of stepping
the clock. This is only applied on the first update. When set to 0.0, the servo
will not step the clock on start.
The default is 0.00002 (20 microseconds).

.TP
.B free_running
When set to 1, no PHC will be adjusted.
This option can be useful in test scenarios, for example to determine
how well synchronized a group of local clocks are to each other.
The default is 0 (adjust the clocks).

.TP
.B leapfile
The path to the current leap seconds definition file. In a Debian
system this file is provided by the tzdata package and can be found at
/usr/share/zoneinfo/leap-seconds.list. If a leapfile is configured it
will be reloaded if modified. The default is an empty string, which
causes the program to use a hard coded table that reflects the known
leap seconds on the date of the software's release.

.TP
.B logging_level
The maximum logging level of messages which should be printed.
The default is 6 (LOG_INFO).

.TP
.B max_frequency
The maximum allowed frequency adjustment of the clock in parts per
billion.  This is an additional limit to the maximum allowed by the
hardware. When set to 0, the hardware limit will be used.
The default is 900000000 (90%).

.TP
.B message_tag
The tag which is added to all messages printed to the standard output
or system log.  If the tag contains the string "{level}", it will be replaced
with the log level of the message as a number.  The default is an empty string
(which cannot be set in the configuration file as the option requires an
argument).

.TP
.B ntpshm_segment
The number of the SHM segment used by ntpshm servo.
The default is 0.

.TP
.B pi_integral_const
The integral constant of the PI controller. When set to 0.0, the
integral constant will be set by the following formula from the current
sync interval.
The default is 0.0.

ki = min(ki_scale * sync^ki_exponent, ki_norm_max / sync)

.TP
.B pi_integral_exponent
The ki_exponent constant in the formula used to set the integral constant of
the PI controller from the sync interval.
The default is 0.4.

.TP
.B pi_integral_norm_max
The ki_norm_max constant in the formula used to set the integral constant of
the PI controller from the sync interval.
The default is 0.3.

.TP
.B pi_integral_scale
The ki_scale constant in the formula used to set the integral constant of
the PI controller from the sync interval.
The default is 0.3.

.TP
.B pi_proportional_const
The proportional constant of the PI controller. When set to 0.0, the
proportional constant will be set by the following formula from the current
sync interval.
The default is 0.0.

kp = min(kp_scale * sync^kp_exponent, kp_norm_max / sync)

.TP
.B pi_proportional_exponent
The kp_exponent constant in the formula used to set the proportional constant of
the PI controller from the sync interval.
The default is \-0.3.

.TP
.B pi_proportional_norm_max
The kp_norm_max constant in the formula used to set the proportional constant of
the PI controller from the sync interval.
The default is 0.7

.TP
.B pi_proportional_scale
The kp_scale constant in the formula used to set the proportional constant of
the PI controller from the sync interval.
The default is 0.7.

.B sa_file
Specifies the location of the file containing Security Associations used
for immediate security processing of the Authentication TLV in support of
the optional security mechanism defined in ieee1588-2019 ch 14.16. See
\fBSECURITY ASSOCIATION OPTIONS\fR for more info on file contents.
The default is an empty string. (disabled).

.TP
.B servo_num_offset_values
The number of offset values considered in order to transition from the
SERVO_LOCKED to the SERVO_LOCKED_STABLE state.
The transition occurs once the last 'servo_num_offset_values' offsets
are all below the 'servo_offset_threshold' value.
The default value is 10.

.TP
.B servo_offset_threshold
The offset threshold used in order to transition from the SERVO_LOCKED
to the SERVO_LOCKED_STABLE state.  The transition occurs once the
last 'servo_num_offset_values' offsets are all below the threshold value.
The default value of offset_threshold is 0 (disabled).

.TP
.B step_threshold
The maximum offset, specified in seconds, that the servo will correct
by changing the clock frequency instead of stepping the clock. When
set to 0.0, the servo will never step the clock except on start.
The default is 0.0.

.TP
.B spp
Specifies the security parameters pointer for the desired security association
to be used for authentication TLV support. If specified, the port owning the
spp will attempt to attach (outbound) and check (inbound) authentication TLVs
for all messages in accordance to the corresponding security association
sourced via the \fBsa_file\fR directive. Not compatible with one step ports.
Must be in the range of -1 to 255, inclusive. The default is -1 (disabled).

.TP
.B ts2phc.external_pps
Enables an external reference with the \fB-a\fP option. If set to 1, ts2phc
will assume the source of the PPS signal is a different clock from the PHCs
used by ptp4l (configured with the \fBboundary_clock_jbod\fP option). The use
case is a holdover using an externally controlled stabilized clock, which is
expected to be synchronized to the PHC that is synchronized by ptp4l, and
running free when ptp4l is not synchronizing any of the PHCs. Note that it is a
different holdover than the one enabled by the \fBts2phc.holdover\fP option
below. The default is 0 (disabled).
.TP
.B ts2phc.holdover
The holdover interval, specified in seconds. When the ToD information stops
working (e.g. GNSS receiver lost its fix), ts2phc is allowed for the specified
interval to continue synchronizing the target clock as long as the servo is in
the SERVO_LOCKED_STABLE state. The servo state needs be enabled by the
\fBservo_offset_threshold\fP option. The holdover is not supported with the
\fB-a\fP option and when \fBts2phc.extts_polarity\fP is set to \fIboth\fP.
The default is 0 (disabled).

.TP
.B ts2phc.nmea_delay
Specifies the minimum expected delay of NMEA RMC messages in nanoseconds.
If the maximum delay is longer than 1 second, or 'ts2phc.pulsewidth'
if 'ts2phc.extts_polarity' is set to "both", this option needs to be set
accordingly to allow the timestamps from NMEA messages to be correctly
assigned to pulses from the PPS signal and wrong PPS edges to be rejected if
the edge rejection is enabled.
The default is 0 nanoseconds.

.TP
.B ts2phc.nmea_remote_host, ts2phc.nmea_remote_port
Specifies the remote host providing ToD information when using the
"nmea" PPS signal source.  Note that if these two options are both
specified, then the given remote connection will be used in preference
to the configured serial port.
These options default to the empty string, that is, not specified.

.TP
.B ts2phc.nmea_serialport, ts2phc.nmea_baudrate
Specifies the serial port and baudrate in bps for character device
providing ToD information when using the "nmea" PPS signal source. Note
that if the options, ts2phc.nmea_remote_host and
ts2phc.nmea_remote_port, are both specified, then the given remote
connection will be used in preference to the configured serial port.
The default serial port is "/dev/ttyS0".
The default baudrate is 9600 bps.

.TP
.B ts2phc.perout_phase
Configures the offset between the beginning of the second and the PPS
source's rising edge. Available only for the PHC kind of PPS source. The supported

ts2phc.c
/**
 * @file ts2phc.c
 * @brief Utility program to synchronize the PHC clock to external events
 * @note Copyright (C) 2013 Balint Ferencz <fernya@sch.bme.hu>
 * @note Based on the phc2sys utility
 * @note Copyright (C) 2012 Richard Cochran <richardcochran@gmail.com>
 * @note SPDX-License-Identifier: GPL-2.0+
 */
#include <errno.h>
#include <net/if.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>

#include "clockadj.h"
#include "config.h"
#include "contain.h"
#include "interface.h"
#include "phc.h"
#include "print.h"
#include "sad.h"
#include "ts2phc.h"
#include "version.h"

#define NS_PER_SEC		1000000000LL
#define SAMPLE_WEIGHT		1.0

struct interface {
	STAILQ_ENTRY(interface) list;
};

static void ts2phc_cleanup(struct ts2phc_private *priv)
{
	struct ts2phc_port *p, *tmp;

	ts2phc_pps_sink_cleanup(priv);
	if (priv->src)
		ts2phc_pps_source_destroy(priv->src);
	if (priv->cfg) {
		sad_destroy(priv->cfg);
		config_destroy(priv->cfg);
	}
	if (priv->agent)
		pmc_agent_destroy(priv->agent);

	/*
	 * Clocks are destroyed by the cleanup methods of the individual
	 * PPS source and sink modules.
	 */
	LIST_FOREACH_SAFE(p, &priv->ports, list, tmp)
		free(p);

	msg_cleanup();
}

static struct ts2phc_port *ts2phc_port_get(struct ts2phc_private *priv,
					   unsigned int number)
{
	struct ts2phc_port *p;

	LIST_FOREACH(p, &priv->ports, list)
		if (p->number == number)
			return p;

	return NULL;
}

static enum port_state ts2phc_clock_compute_state(struct ts2phc_private *priv,
						  struct ts2phc_clock *clock)
{
	enum port_state state = PS_DISABLED;
	struct ts2phc_port *p;

	LIST_FOREACH(p, &priv->ports, list) {
		if (p->clock != clock)
			continue;
		/* PS_SLAVE takes the highest precedence, PS_UNCALIBRATED
		 * after that, PS_MASTER is third, PS_PRE_MASTER fourth and
		 * all of that overrides PS_DISABLED, which corresponds
		 * nicely with the numerical values */
		if (p->state > state)
			state = p->state;
	}
	return state;
}

static int ts2phc_recv_subscribed(void *context, struct ptp_message *msg,
				  int excluded)
{
	struct ts2phc_private *priv = context;
	enum port_state state;
	struct ts2phc_clock *clock;
	struct portDS *pds;
	struct ts2phc_port *port;
	int mgt_id;

	mgt_id = management_tlv_id(msg);
	if (mgt_id == excluded)
		return 0;

	switch (mgt_id) {
	case MID_PORT_DATA_SET:
		pds = management_tlv_data(msg);
		port = ts2phc_port_get(priv, pds->portIdentity.portNumber);
		if (!port) {
			pr_info("received data for unknown port %s",
				pid2str(&pds->portIdentity));
			return 1;
		}
		state = port_state_normalize(pds->portState);
		if (port->state != state) {
			pr_info("port %s changed state",
				pid2str(&pds->portIdentity));
			port->state = state;
			clock = port->clock;
			state = ts2phc_clock_compute_state(priv, clock);
			if (clock->state != state || clock->new_state) {
				clock->new_state = state;
				priv->state_changed = true;
			}
		}
		return 1;
	}
	return 0;
}

static struct servo *ts2phc_servo_create(struct ts2phc_private *priv,
					 struct ts2phc_clock *clock)
{
	enum servo_type type = config_get_int(priv->cfg, NULL, "clock_servo");
	struct servo *servo;
	double fadj;
	int max_adj;

	fadj = clockadj_get_freq(clock->clkid);

	max_adj = phc_max_adj(clock->clkid);

	servo = servo_create(priv->cfg, type, -fadj, max_adj, 0);
	if (!servo)
		return NULL;

	servo_sync_interval(servo, priv->pulse_period / 1.0e9);

	return servo;
}

void ts2phc_clock_add_tstamp(struct ts2phc_clock *clock, tmv_t t)
{
	struct timespec ts = tmv_to_timespec(t);

	pr_debug("adding tstamp %ld.%09ld to clock %s",
		 ts.tv_sec, ts.tv_nsec, clock->name);
	clock->last_ts = t;
	clock->is_ts_available = true;
}

static int ts2phc_clock_get_tstamp(struct ts2phc_clock *clock, tmv_t *ts)
{
	if (!clock->is_ts_available)
		return 0;
	clock->is_ts_available = false;
	*ts = clock->last_ts;
	return 1;
}

static void ts2phc_clock_flush_tstamp(struct ts2phc_clock *clock)
{
	clock->is_ts_available = false;
}

struct ts2phc_clock *ts2phc_clock_add(struct ts2phc_private *priv,
				      const char *device)
{
	clockid_t clkid = CLOCK_INVALID;
	struct ts2phc_clock *c;
	int phc_index = -1;
	int err;

	clkid = posix_clock_open(device, &phc_index);
	if (clkid == CLOCK_INVALID)
		return NULL;

	LIST_FOREACH(c, &priv->clocks, list) {
		if (c->phc_index == phc_index) {
			/* Already have the clock, don't add it again */
			posix_clock_close(clkid);
			return c;
		}
	}

	c = calloc(1, sizeof(*c));
	if (!c) {
		pr_err("failed to allocate memory for a clock");
		return NULL;
	}
	c->clkid = clkid;
	c->fd = CLOCKID_TO_FD(clkid);
	c->phc_index = phc_index;
	c->servo_state = SERVO_UNLOCKED;
	c->servo = ts2phc_servo_create(priv, c);
	c->no_adj = config_get_int(priv->cfg, NULL, "free_running");
	err = asprintf(&c->name, "/dev/ptp%d", phc_index);
	if (err < 0) {
		free(c);
		posix_clock_close(clkid);
		return NULL;
	}

	LIST_INSERT_HEAD(&priv->clocks, c, list);
	return c;
}

void ts2phc_clock_destroy(struct ts2phc_clock *c)
{
	servo_destroy(c->servo);
	posix_clock_close(c->clkid);
	free(c->name);
	free(c);
}

static struct ts2phc_port *ts2phc_port_add(struct ts2phc_private *priv,
					   unsigned int number, char *device)
{
	struct ts2phc_clock *c = NULL;
	struct ts2phc_port *p, *tmp;

	p = ts2phc_port_get(priv, number);
	if (p)
		return p;
	/* port is a new one, look whether we have the device already
	 * on a different port
	 */
	LIST_FOREACH(tmp, &priv->ports, list) {
		if (tmp->number == number) {
			c = tmp->clock;
			break;
		}
	}
	if (!c) {
		c = ts2phc_clock_add(priv, device);
		if (!c)
			return NULL;
	}
	p = malloc(sizeof(*p));
	if (!p) {
		pr_err("failed to allocate memory for a port");
		ts2phc_clock_destroy(c);
		return NULL;
	}
	p->number = number;
	p->clock = c;
	LIST_INSERT_HEAD(&priv->ports, p, list);
	return p;
}

static int ts2phc_auto_init_ports(struct ts2phc_private *priv)
{
	int number_ports, timestamping, phc_index, err;
	struct ts2phc_clock *clock;
	struct ts2phc_port *port;
	enum port_state state;
	char iface[IFNAMSIZ];
	unsigned int i;

	while (1) {
		if (!is_running())
			return -1;
		err = pmc_agent_query_dds(priv->agent, 1000);
		if (!err)
			break;
		if (err == -ETIMEDOUT)
			pr_notice("Waiting for ptp4l...");
		else
			return -1;
	}

	number_ports = pmc_agent_get_number_ports(priv->agent);
	if (number_ports <= 0) {
		pr_err("failed to get number of ports");
		return -1;
	}

	err = pmc_agent_subscribe(priv->agent, 1000, 1);
	if (err) {
		pr_err("failed to subscribe");
		return -1;
	}

	for (i = 1; i <= number_ports; i++) {
		err = pmc_agent_query_port_properties(priv->agent, 1000, i,
						      &state, &timestamping,
						      &phc_index, iface);
		if (err == -ENODEV) {
			/* port does not exist, ignore the port */
			continue;
		}
		if (err) {
			pr_err("failed to get port properties");
			return -1;
		}
		if (timestamping == TS_SOFTWARE) {
			/* ignore ports with software time stamping */
			continue;
		}
		port = ts2phc_port_add(priv, i, iface);
		if (!port)
			return -1;
		port->state = port_state_normalize(state);
	}
	if (LIST_EMPTY(&priv->clocks)) {
		pr_err("no suitable ports available");
		return -1;
	}
	LIST_FOREACH(clock, &priv->clocks, list) {
		clock->new_state = ts2phc_clock_compute_state(priv, clock);
	}
	priv->state_changed = true;

	return 0;
}

static void ts2phc_reconfigure(struct ts2phc_private *priv)
{
	struct ts2phc_clock *c, *ref_clk = NULL, *last = NULL;
	int num_ref_clocks = 0, num_target_clocks = 0;

	pr_info("reconfiguring after port state change");
	priv->state_changed = false;

	LIST_FOREACH(c, &priv->clocks, list) {
		if (c->new_state) {
			c->state = c->new_state;
			c->new_state = 0;
		}

		switch (c->state) {
		case PS_FAULTY:
		case PS_DISABLED:
		case PS_LISTENING:
		case PS_PRE_MASTER:
		case PS_MASTER:
		case PS_PASSIVE:
			if (!c->is_target) {
				pr_info("selecting %s for synchronization",
					c->name);
				c->is_target = true;
			}
			num_target_clocks++;
			break;
		case PS_SLAVE:
			ref_clk = c;
			/* Fall through */
		case PS_UNCALIBRATED:
			num_ref_clocks++;
			if (priv->external_pps && c->is_target) {
				pr_info("unselecting %s for synchronization",
					c->name);
				c->is_target = false;
			}
			break;
		default:
			break;
		}
		last = c;
	}
	if (priv->external_pps) {
		pr_info("assuming external reference clock");
		return;
	}
	if (num_target_clocks >= 1 && !ref_clk) {
		priv->ref_clock = last;
		priv->ref_clock->is_target = false;
		/* Reset to original state in next reconfiguration. */
		priv->ref_clock->new_state = priv->ref_clock->state;
		priv->ref_clock->state = PS_SLAVE;
		pr_info("no reference clock, selecting %s by default",
			last->name);
		return;
	}
	if (num_ref_clocks > 1) {
		pr_info("multiple reference clocks available, postponing sync...");
		priv->ref_clock = NULL;
		return;
	}
	if (num_ref_clocks > 0 && !ref_clk) {
		pr_info("reference clock not ready, waiting...");
		priv->ref_clock = NULL;
		return;
	}
	if (!num_ref_clocks && !num_target_clocks) {
		pr_info("no PHC ready, waiting...");
		priv->ref_clock = NULL;
		return;
	}
	if (!num_ref_clocks) {
		pr_info("nothing to synchronize");
		priv->ref_clock = NULL;
		return;
	}
	ref_clk->is_target = false;
	priv->ref_clock = ref_clk;
	pr_info("selecting %s as the reference clock", ref_clk->name);
}

static int ts2phc_pps_source_implicit_tstamp(struct ts2phc_private *priv,
					     tmv_t *source_tmv)
{
	struct timespec source_ts;
	tmv_t tmv;
	int err;

	err = ts2phc_pps_source_getppstime(priv->src, &source_ts);
	if (err < 0) {
		pr_err("source ts not valid");
		return err;
	}

	tmv = timespec_to_tmv(source_ts);
	tmv = tmv_sub(tmv, priv->perout_phase);
	source_ts = tmv_to_timespec(tmv);

	/*
	 * As long as the kernel doesn't support a proper API for reporting
	 * back a precise perout timestamp, we'll have to implicitly assume
	 * assumption that the current time on the PPS source is still within
	 * +/- half a second of the past perout output edge, and hence, we can
	 * deduce the timestamp (actually only seconds part, nanoseconds are by
	 * construction zero) of this edge at the emitter based on the
	 * emitter's current time.
	 *
	 * With an NMEA source assume its messages always follow the pulse, i.e.
	 * assign the timestamp to the previous pulse instead of nearest pulse.
	 *
	 * To support >1Hz PPS signals, work with the 1/rate period instead of
	 * second.
	 */
	if (ts2phc_pps_source_get_type(priv->src) == TS2PHC_PPS_SOURCE_NMEA) {
		source_ts.tv_nsec += priv->pulse_period;
	} else {
		source_ts.tv_nsec += priv->pulse_period / 2;
	}

	/* Truncate the timestamp to the pulse period boundary */
	source_ts.tv_nsec /= priv->pulse_period;
	source_ts.tv_nsec *= priv->pulse_period;
	if (source_ts.tv_nsec >= NS_PER_SEC) {
		source_ts.tv_sec++;
		source_ts.tv_nsec -= NS_PER_SEC;
	}

	tmv = timespec_to_tmv(source_ts);
	tmv = tmv_add(tmv, priv->perout_phase);

	*source_tmv = tmv;

	return 0;
}

static void ts2phc_synchronize_clocks(struct ts2phc_private *priv, int autocfg)
{
	struct timespec source_ts, now;
	tmv_t source_tmv;
	struct ts2phc_clock *c;
	int holdover, valid;

	if (autocfg && !priv->external_pps) {
		if (!priv->ref_clock) {
			pr_debug("no reference clock, skipping");
			return;
		}
		valid = ts2phc_clock_get_tstamp(priv->ref_clock, &source_tmv);
		if (!valid) {
			pr_err("reference clock (%s) timestamp not valid, skipping",
				priv->ref_clock->name);
			return;
		}
	} else {
		valid = !ts2phc_pps_source_implicit_tstamp(priv, &source_tmv);
	}

	if (valid) {
		priv->holdover_start = 0;
		holdover = 0;
	} else {
		clock_gettime(CLOCK_MONOTONIC, &now);

		if (!priv->holdover_start)
			priv->holdover_start = now.tv_sec;
		if (now.tv_sec >= priv->holdover_start + priv->holdover_length)
			return;
		holdover = 1;
	}

	LIST_FOREACH(c, &priv->clocks, list) {
		int64_t offset;
		double adj;
		tmv_t ts;

		if (!c->is_target)
			continue;

		valid = ts2phc_clock_get_tstamp(c, &ts);
		if (!valid) {
			pr_debug("%s timestamp not valid, skipping", c->name);
			continue;
		}

		if (holdover) {
			if (c->servo_state != SERVO_LOCKED_STABLE)
				continue;
			source_ts = tmv_to_timespec(ts);
			if (source_ts.tv_nsec > NS_PER_SEC / 2)
				source_ts.tv_sec++;
			source_ts.tv_nsec = 0;
			source_tmv = timespec_to_tmv(source_ts);
		}

		offset = tmv_to_nanoseconds(tmv_sub(ts, source_tmv));

		if (c->no_adj) {
			pr_info("%s offset %10" PRId64, c->name,
				offset);
			continue;
		}

		adj = servo_sample(c->servo, offset, tmv_to_nanoseconds(ts),
				   SAMPLE_WEIGHT, &c->servo_state);

		if (holdover && c->servo_state != SERVO_LOCKED_STABLE) {
			pr_info("%s lost holdover lock (offset %10" PRId64 ")",
				c->name, offset);
			continue;
		}

		pr_info("%s offset %10" PRId64 " s%d freq %+7.0f%s",
			c->name, offset, c->servo_state, adj,
			holdover ? " holdover" : "");

		switch (c->servo_state) {
		case SERVO_UNLOCKED:
			break;
		case SERVO_JUMP:
			if (0 != clockadj_set_freq(c->clkid, -adj) &&
			    0 != clockadj_step(c->clkid, -offset)) {
				goto servo_unlock;
			}
			break;
		case SERVO_LOCKED:
		case SERVO_LOCKED_STABLE:
			if (clockadj_set_freq(c->clkid, -adj)) {
				goto servo_unlock;
			}
			break;
		}
		continue;

servo_unlock:
		servo_reset(c->servo);
		c->servo_state = SERVO_UNLOCKED;
	}
}

static int ts2phc_collect_pps_source_tstamp(struct ts2phc_private *priv)
{
	struct ts2phc_clock *pps_src_clock;
	tmv_t source_tmv;
	int err;

	pps_src_clock = ts2phc_pps_source_get_clock(priv->src);
	/*
	 * PPS source isn't a PHC (it may be a generic or a GPS PPS source),
	 * don't error out, just don't do anything. If it doesn't have a PHC,
	 * there is nothing to synchronize, which is the only point of
	 * collecting its perout timestamp in the first place.
	 */
	if (!pps_src_clock)
		return 0;

	err = ts2phc_pps_source_implicit_tstamp(priv, &source_tmv);
	if (err < 0)
		return err;

	ts2phc_clock_add_tstamp(pps_src_clock, source_tmv);

	return 0;
}

static void usage(char *progname)
{
	fprintf(stderr,
		"\n"
		"usage: %s [options]\n\n"
		" -a             turn on autoconfiguration\n"
		" -c [dev|name]  PHC time sink (like /dev/ptp0 or eth0)\n"
		"                (may be specified multiple times)\n"
		" -f [file]      read configuration from 'file'\n"
		" -h             prints this message and exits\n"
		" -l [num]       set the logging level to 'num'\n"
		" -m             print messages to stdout\n"
		" -q             do not print messages to the syslog\n"
		" -s [dev|name]  source of the PPS signal\n"
		"                may take any of the following forms:\n"
		"                    generic   - an external 1-PPS without ToD information\n"
		"                    /dev/ptp0 - a local PTP Hardware Clock (PHC)\n"
		"                    eth0      - a local PTP Hardware Clock (PHC)\n"
		"                    nmea      - a gps device connected by serial port or network\n"
		" -v             prints the software version and exits\n"
		"\n",
		progname);
}

int main(int argc, char *argv[])
{
	int c, err = 0, have_sink = 0, index, cmd_line_print_level;
	char uds_local[MAX_IFNAME_SIZE + 1];
	enum ts2phc_pps_source_type pps_type;
	struct ts2phc_private priv = {0};
	char *config = NULL, *progname;
	const char *tod_source = NULL;
	struct config *cfg = NULL;
	struct interface *iface;
	struct option *opts;
	int autocfg = 0;

	handle_term_signals();

	cfg = config_create();
	if (!cfg) {
		ts2phc_cleanup(&priv);
		return -1;
	}
	priv.cfg = cfg;
	priv.agent = pmc_agent_create();
	if (!priv.agent) {
		ts2phc_cleanup(&priv);
		return -1;
	}

	opts = config_long_options(cfg);

	/* Process the command line arguments. */
	progname = strrchr(argv[0], '/');
	progname = progname ? 1 + progname : argv[0];
	while (EOF != (c = getopt_long(argc, argv, "ac:f:hi:l:mqs:v", opts, &index))) {
		switch (c) {
		case 0:
			if (config_parse_option(cfg, opts[index].name, optarg)) {
				ts2phc_cleanup(&priv);
				return -1;
			}
			break;
		case 'a':
			autocfg = 1;
			break;
		case 'c':
			if (!config_create_interface(optarg, cfg)) {
				fprintf(stderr, "failed to add PPS sink\n");
				ts2phc_cleanup(&priv);
				return -1;
			}
			have_sink = 1;
			break;
		case 'f':
			config = optarg;
			break;
		case 'l':
			if (get_arg_val_i(c, optarg, &cmd_line_print_level,
					  PRINT_LEVEL_MIN, PRINT_LEVEL_MAX)) {
				ts2phc_cleanup(&priv);
				return -1;
			}
			config_set_int(cfg, "logging_level", cmd_line_print_level);
			break;
		case 'm':
			config_set_int(cfg, "verbose", 1);
			print_set_verbose(1);
			break;
		case 'q':
			config_set_int(cfg, "use_syslog", 0);
			print_set_syslog(0);
			break;
		case 's':
			if (tod_source) {
				fprintf(stderr, "too many PPS sources\n");
				ts2phc_cleanup(&priv);
				return -1;
			}
			tod_source = optarg;
			break;
		case 'v':
			ts2phc_cleanup(&priv);
			version_show(stdout);
			return 0;
		case 'h':
			ts2phc_cleanup(&priv);
			usage(progname);
			return 0;
		case '?':
		default:
			ts2phc_cleanup(&priv);
			usage(progname);
			return -1;
		}
	}
	if (config && (c = config_read(config, cfg))) {
		fprintf(stderr, "failed to read config\n");
		ts2phc_cleanup(&priv);
		return -1;
	}
	print_set_progname(progname);
	print_set_tag(config_get_string(cfg, NULL, "message_tag"));
	print_set_verbose(config_get_int(cfg, NULL, "verbose"));
	print_set_syslog(config_get_int(cfg, NULL, "use_syslog"));
	print_set_level(config_get_int(cfg, NULL, "logging_level"));

	/* Set this variable early for added sinks */
	priv.pulse_period = NSEC_PER_SEC / config_get_int(cfg, NULL,
							  "ts2phc.pulserate");

	STAILQ_INIT(&priv.sinks);

	if (sad_create(cfg)) {
		fprintf(stderr, "failed to get security associations\n");
		ts2phc_cleanup(&priv);
		return -1;
	}

	snprintf(uds_local, sizeof(uds_local), "ts2phc.%d", getpid());

	if (autocfg) {
		err = init_pmc_node(cfg, priv.agent, uds_local,
				    ts2phc_recv_subscribed, &priv);
		if (err) {
			ts2phc_cleanup(&priv);
			return -1;
		}
		err = ts2phc_auto_init_ports(&priv);
		if (err) {
			ts2phc_cleanup(&priv);
			return -1;
		}
	}

	STAILQ_FOREACH(iface, &cfg->interfaces, list) {
		const char *dev = interface_name(iface);

		if (1 == config_get_int(cfg, dev, "ts2phc.master")) {
			int perout_phase;

			if (tod_source) {
				fprintf(stderr, "too many PPS sources\n");
				ts2phc_cleanup(&priv);
				return -1;
			}
			tod_source = dev;
			perout_phase = config_get_int(cfg, dev,
						      "ts2phc.perout_phase");
			/*
			 * We use a default value of -1 to distinguish whether
			 * to use the PTP_PEROUT_PHASE API or not. But if we
			 * don't use that (and therefore we use absolute start
			 * time), the phase is still zero, by our application's
			 * convention.
			 */
			if (perout_phase < 0)
				perout_phase = 0;
			priv.perout_phase = nanoseconds_to_tmv(perout_phase);
		} else {
			if (ts2phc_pps_sink_add(&priv, interface_name(iface))) {
				fprintf(stderr, "failed to add PPS sink\n");
				ts2phc_cleanup(&priv);
				return -1;
			}
			have_sink = 1;
		}
	}
	if (!have_sink) {
		fprintf(stderr, "no PPS sinks specified\n");
		ts2phc_cleanup(&priv);
		usage(progname);
		return -1;
	}

	if (!tod_source)
		tod_source = config_get_string(cfg, NULL, "ts2phc.tod_source");

	if (ts2phc_pps_sinks_init(&priv)) {
		fprintf(stderr, "failed to initialize PPS sinks\n");
		ts2phc_cleanup(&priv);
		return -1;
	}

	if (!strcasecmp(tod_source, "generic")) {
		pps_type = TS2PHC_PPS_SOURCE_GENERIC;
	} else if (!strcasecmp(tod_source, "nmea")) {
		pps_type = TS2PHC_PPS_SOURCE_NMEA;
	} else {
		pps_type = TS2PHC_PPS_SOURCE_PHC;
	}
	priv.src = ts2phc_pps_source_create(&priv, tod_source, pps_type);
	if (!priv.src) {
		fprintf(stderr, "failed to create PPS source\n");
		ts2phc_cleanup(&priv);
		return -1;
	}

	priv.external_pps = config_get_int(cfg, NULL, "ts2phc.external_pps");

	priv.holdover_length = config_get_int(cfg, NULL, "ts2phc.holdover");
	priv.holdover_start = 0;

	while (is_running()) {
		struct ts2phc_clock *clk;

		if (autocfg) {
			/* Collect updates from ptp4l */
			err = pmc_agent_update(priv.agent);
			if (err < 0) {
				pr_err("pmc_agent_update returned %d", err);
				break;
			}

			if (priv.state_changed)
				ts2phc_reconfigure(&priv);
		}

		LIST_FOREACH(clk, &priv.clocks, list)
			ts2phc_clock_flush_tstamp(clk);

		err = ts2phc_pps_sink_poll(&priv);
		if (err < 0) {
			pr_err("poll failed");
			break;
		}
		if (err > 0) {
			err = ts2phc_collect_pps_source_tstamp(&priv);
			if (err) {
				pr_err("failed to collect PPS source tstamp");
				break;
			}

			ts2phc_synchronize_clocks(&priv, autocfg);
		}
	}

	ts2phc_cleanup(&priv);
	return err;
}
